frontend/src/flashcard/flashcardView.js

```diff
@@ -0,0 +1,375 @@
+            starredCards,
+            cardIndex,
+            cardData,
+        } = this.state;
+        return starOnly
+            ? cardData[starredCards[cardIndex]]
+            : cardData[cardIndex];
+    }
+
+    changeCard = (delta) => {
+        const { cardData, starredCards, cardIndex } = this.state;
+        const listLength = this.state.starOnly ? starredCards.length : cardData.length;
+        const newCardIndex = (cardIndex + delta + listLength) % listLength;
+        this.setState({
+            showBack: false,
+            cardIndex: newCardIndex,
+            showNext: true,
+        });
+    }
+
+    flipCard = () => {
+        this.setState({
+            showBack: !this.state.showBack,
+            showNext: false,
+        });
+    }
+
+    isStarred = (cardIndex) => {
+        const index = this.state.starredCards.indexOf(cardIndex);
+        return index !== -1;
+    }
+
+    toggleStar = () => {
+        const { starredCards, starOnly, cardIndex } = this.state;
+        let currentIndex = starOnly ? starredCards[cardIndex] : cardIndex;
+        if (this.isStarred(currentIndex)) {
+            starredCards.splice(starredCards.indexOf(currentIndex), 1);
+            if (starOnly) {
+                if (starredCards.length > 0) {
+                    currentIndex %= starredCards.length;
+                } else {
+                    currentIndex = 0;
+                }
+            }
+        } else {
+            starredCards.push(this.state.cardIndex);
+        }
+        this.setState({ starredCards, cardIndex: currentIndex });
+    }
+
+    toggleStarOnly = () => {
+        this.setState({ cardIndex: 0, starOnly: !this.state.starOnly });
+    }
+
+    handleKeyDown = (e) => {
+        if (['ArrowLeft', 'ArrowRight', 'Space', 'KeyS', 'KeyO'].includes(e.code)) {
+            e.preventDefault();
+            const { code } = e;
+            if (code === 'ArrowLeft') {
+                this.changeCard(-1);
+            } else if (code === 'ArrowRight') {
+                this.changeCard(1);
+            } else if (code === 'Space') {
+                this.flipCard();
+            } else if (code === 'KeyS') {
+                this.toggleStar();
+            } else if (code === 'KeyO') {
+                this.toggleStarOnly();
+            }
+        }
+    }
+
+    modalHandler = (event) => {
+        event.preventDefault();
+        this.setState({
+            showModal: !this.state.showModal,
+        });
+    }
+
+    checkProgress = () => {
+        const numStarred = this.state.starredCards.length;
+        const numCards = this.state.cardData.length;
+        const currentCard = this.state.cardIndex + 1;
+        if (this.state.starOnly && numStarred !== 0) {
+            return parseInt((currentCard / numStarred) * 100);
+        }
+        if (!this.state.starOnly && numCards !== 0) {
+            return parseInt((currentCard / numCards) * 100);
+        }
+        return 0;
+    }
+
+    render() {
```

Comment:
The long chunks of JSX in this function are difficult to read. For maintainability, better to break them down into smaller chunks. Just like you pulled `flipcard` out into its own chunk, you could pull the progress bar, the instructions modal, etc. into individual functional components that just take whatever props they need to render. My preference is usually to pull these _out_ of the render function into helper methods on the class like `renderFlipcard(whateverPropsItNeeds`, `renderHelpModal(whateverProps)`, that take in the necessary props and output some JSX.

Then, the JSX in your render call becomes simpler:

e.g., something like
```
return (
    <Navbar/>
    <div className='page' onKeyDown={this.handleKeyDown}>
        {renderHeader()}
        {renderHelpModal()}
        // etc. etc.
    </div>
)
```

For a big container component like this, separating the view logic for individual bits like that will help keep the thing maintainable.